server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceFieldConfig.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.search.MultiValueMode;
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.function.BiFunction;
+
+public class MultiValuesSourceFieldConfig implements Writeable, ToXContentFragment {
```

Comment:
Could this maybe wrap `ValuesSourceConfig` so we ensure as feeatures are added to one they are added to the other?

Reply:
So the layout is a bit tricky, and naming is maybe (probably) confusing.  Open to suggestions.

`MultiValuesSourceFieldConfig` is spiritually related to `ValuesSourceParseHelper#declareFields()` in that it is basically the parser and builder object for the commonly shared fields

I think `MultiValuesSourceConfig` is closer to what you were expecting, which is the final object.  This contains a map of fields, where each entry's value is a Wrapper object containing a `ValuesSourceConfig` and a `MultiValueMode`.

So the underlying features should be shared, but the parsing is indeed still different.  I can see if there's a way to share the parsing, but it may be tricky since the regular ValuesSourceConfig also defines a `targetValueType` as part of the common fields, but that only applies to the total MultiValuesSource, not the individual fields.

I'll poke at it a bit.  Might be easier to zoom about this when you're back too.